src/backend/utils/misc/guc_gp.c

```diff
@@ -5167,7 +5167,7 @@ check_optimizer(bool *newval, void **extra, GucSource source)
 		 */
 		if (GP_ROLE_DISPATCH != Gp_role)
 			*newval = false;
-		else if (NULL == planner_hook)
+		else if (source > PGC_S_ARGV && NULL == planner_hook)
```

Comment:
planner_hook is checked twice, why not something
```suggestion
		else if (NULL == planner_hook)
		{
			if (source > PGC_S_ARGV)
			{
				GUC_check_errmsg("External planner is not registered");
				return false;
			}

			elog(LOG, "'optimizer' is explicitly set to 'on'. "
				"Please ensure that an external planner is added to 'shared_preload_libraries'");
		}
```